### PR TITLE
Added is_move_legal to de-duplicate logic

### PR DESCRIPTION
**Pull request by sethtroisi**
Cleanup some duplicate logic in FastState, Training, and UCTNode

Tested by verifying several training.gz files matched sha1sum and looking at some ko positions.

`$  echo "dump_supervised all.sgf training.data" | ./leelaz -s 123 -w ../weights.txt`
`$ sha1sum training.data.*`

### CODE REVIEW

File: src/UCTNode.cpp [hunk 1]
```diff
@@ -97,31 +97,24 @@ bool UCTNode::create_children(std::atomic<int> & nodecount,
...
     }
     eval = net_eval;
 
-    FastBoard & board = state.board;
     std::vector<Network::scored_node> nodelist;
 
     auto legal_sum = 0.0f;
     for (auto& node : raw_netlist.first) {
         auto vertex = node.second;
-        if (vertex != FastBoard::PASS) {
```
Comment (sethtroisi) [inline src/UCTNode.cpp]: I'm not sure where this checks that the vertex isn't occupied, probably a side effect of `!board.is_suicide` but I didn't find an obvious check.

Either way fastfailing if square is occupied (in is_move_legal) is probably better

Comment (gcp) [reply src/UCTNode.cpp]: https://github.com/gcp/leela-zero/blob/1894e17d8165b8f88b03f704155400091c07db7b/src/Network.cpp#L459

**Comment by gcp** [comment]
Needs a rebase.

**Comment by sethtroisi** [comment]
Rebased

**Pull request merged**